consensus/consensus-types/src/common.rs

```diff
@@ -233,18 +279,90 @@ impl Payload {
             Payload::InQuorumStore(proof_with_status) => proof_with_status.proofs.is_empty(),
             Payload::InQuorumStoreWithLimit(proof_with_status) => {
                 proof_with_status.proof_with_data.proofs.is_empty()
-                    || (proof_with_status.max_txns_to_execute.is_some()
-                        && proof_with_status.max_txns_to_execute.unwrap() == 0)
+                    || proof_with_status.max_txns_to_execute == Some(0)
+            },
+            Payload::QuorumStoreInlineHybrid(
+                inline_batches,
+                proof_with_data,
+                max_txns_to_execute,
+            ) => {
+                *max_txns_to_execute == Some(0)
+                    || (proof_with_data.proofs.is_empty() && inline_batches.is_empty())
             },
         }
     }
 
-    pub fn extend(&mut self, other: Payload) {
+    pub fn extend(self, other: Payload) -> Self {
         match (self, other) {
-            (Payload::DirectMempool(v1), Payload::DirectMempool(v2)) => v1.extend(v2),
-            (Payload::InQuorumStore(p1), Payload::InQuorumStore(p2)) => p1.extend(p2),
+            (Payload::DirectMempool(v1), Payload::DirectMempool(v2)) => {
+                let mut v3 = v1;
+                v3.extend(v2);
+                Payload::DirectMempool(v3)
+            },
+            (Payload::InQuorumStore(p1), Payload::InQuorumStore(p2)) => {
+                let mut p3 = p1;
+                p3.extend(p2);
+                Payload::InQuorumStore(p3)
+            },
             (Payload::InQuorumStoreWithLimit(p1), Payload::InQuorumStoreWithLimit(p2)) => {
-                p1.extend(p2)
+                let mut p3 = p1;
+                p3.extend(p2);
+                Payload::InQuorumStoreWithLimit(p3)
+            },
+            (
+                Payload::QuorumStoreInlineHybrid(b1, p1, m1),
+                Payload::QuorumStoreInlineHybrid(b2, p2, m2),
+            ) => {
+                let mut b3 = b1;
+                b3.extend(b2);
+                let mut p3 = p1;
+                p3.extend(p2);
+                // TODO: What's the right logic here?
+                let m3 = if m1.is_none() {
```

Comment:
nit: Can you change this to `match` statement? It would be more readable and cleaner that way rather than a bunch of if-else. Something like this should do 

```
match (m1, m2) {
 None,  m2 => m2,
 Some(m1), Some(m2) =>
.......

```

Reply:
Done.